megatron/data/dataset_utils.py

```diff
@@ -231,24 +247,40 @@ def create_masked_lm_predictions(tokens,
     if masked_lm_prob == 0:
         return (output_tokens, masked_lm_positions,
                 masked_lm_labels, token_boundary)
+    if prefix_lm:
+        # Adjust probabilities so that the mean is centered at the
+        # correct position.
+        # If we do not do this, the mean is at
+        # `len(tokens) * masked_lm_prob / 2`.
+        masked_lm_prob *= 2
```

Comment:
I re-used the `max_predictions_per_seq` line from the `T5Dataset`, since I needed a trustworthy value for this while handling the limits of the implementation. As you say, `created_masked_lm_predictions` does not handle an arbitrary amount of maximum predictions, so I had to supply something.

I agree about your points here, though: Since we are using a non-capped distribution for the R- and X-denoisers, the implementation does not correctly match what we want. However, both mean and standard deviation for the amount of masked tokens are still very close to the `seqio` paper implementation. As a matter of fact, both are slightly higher for this implementation, contrary to what you would expect given that we have an upper bound. I attributed this to a difference in implementation of the masking code.

The whole `created_masked_lm_predictions` function really is very convoluted already – trying to make it a jack of all trades was likely a wrong decision.

Reply:
While I have not decided to implement the function in separate yet, this hack is now removed. The problem was the further below snippet appending to `ngram_indexes`. In addition, `max_predictions_per_seq` is now fixed for the PrefixLM task. We keep it the same for the other tasks as this is not specified further in the UL2 paper. The normal/uniform sampling only concerns the length of the spans.